### ISSUE TITLE
Op arguments don't get passed in when using .collect() upstream if the op is executed in Databricks

### ISSUE THREAD

**Issue by andrew-arkhipov** [dagster-io/dagster#6853]
## Summary
When using `.collect()` downstream from a dynamically generated op, the resulting values that should populate the list that contains the outputs disappear if the subsequent (downstream) op that consumes said resultant list is executed in Databricks.

## Reproduction
```python
import pkg_resources
from pathlib import Path
from typing import Any, Dict, List

from dagster import DynamicOut, DynamicOutput, In, graph, op
from dagster_databricks import databricks_pyspark_step_launcher

LOCAL_PIPELINE_PACKAGE_PATH = str(Path(pkg_resources.resource_filename("<PARENT_DIR>", "")).parent)


def get_databricks_pyspark_step_launcher(cluster_spec: Dict[str, Any]):
    if not cluster_spec:
        raise ValueError("Expected non-empty cluster spec!")
    
    step_launcher = databricks_pyspark_step_launcher.configured(
        {
            "run_config": {
                "run_name": "databricks job run",
                "cluster": cluster_spec,
                "libraries": [
                    {"pypi": {"package": "dagster-aws==0.13.14"}},
                ]
            },
            "databricks_host": {"env": "DATABRICKS_HOST"},
            "databricks_token": {"env": "DATABRICKS_TOKEN"},
            "local_pipeline_package_path": LOCAL_PIPELINE_PACKAGE_PATH,
            "secrets_to_env_variables": [],
            "storage": {
                "s3": {
                    "access_key_key": "AWS_ACCESS_KEY_ID",
                    "secret_key_key":"AWS_SECRET_ACCESS_KEY",
                    "secret_scope": "dummy",
                }
            },
            "env_variables": {  # this is for a personal use-case -> side note: would it be possible to implement something like this in dagster_databricks?
                "DAGSTER_ENV": "prd",
            },
        }
    )
    return step_launcher


@op(
    config_schema={"num_tasks": int},
    out=DynamicOut(int),
)
def generate(context):
    for i in range(context.op_config["num_tasks"]):
        yield DynamicOutput(value=i, mapping_key=str(i))


@op(ins={"task_num": In(int)})
def add_five(context, task_num):
    return task_num + 5


@op(
    ins={"all_nums": In(List[int])},
    required_resource_keys={"pyspark_step_launcher"},
)
def display(context, all_nums):
    context.log.info(f"GOT NUMS: {all_nums}")


@graph
def example():
    tasks = generate()
    nums = tasks.map(add_five)
    display(nums.collect())


example_job = example.to_job(
    name="example",
    resource_defs={
        "pyspark_step_launcher": get_pyspark_step_launcher(
            cluster_spec={"existing": "<DATABRICKS_CLUSTER_ID>"},
        ),
    },
)
```
We should see `"GOT NUMS: [5, 6, 7]"` if the initial `num_tasks` config parameter is set to 3. Instead, we get `"GOT NUMS: []"`.

## Additional Info about Your Environment
This issue doesn't occur locally - only when the final `display` op is executed in Databricks. Furthermore, version `0.3.14` is being used for the entire suite of Dagster packages.

---
#### Message from the maintainers:
Impacted by this bug? Give it a 👍. We factor engagement into prioritization.

**Comment by andrew-arkhipov** [comment]
Hey Owen, thanks for taking a look at this. I've tried this out with a custom build, and there are a few remaining issues. 

Firstly, it seems like we need to define `__getnewargs__` in the `StepRunRef` class seen [here](https://github.com/dagster-io/dagster/blob/owen/fix-dynamic-output-step-launcher/python_modules/dagster/dagster/core/definitions/step_launcher.py#L13). I couldn't get the fix to work without adding that method (see an example [here](https://stackoverflow.com/questions/1978264/what-does-getnewargs-do-in-this-code)).

Secondly, we're using version `0.13.14` in our environment, and can't afford to update to a newer version yet. Would it be possible for you to create a fix/build for this that uses `0.13.14` that we can use?

Thanks again.

**Comment by andrew-arkhipov** [comment]
@OwenKephart I didn't see `__getnewargs__` added to the PR, were you able to get it working without that?

**Comment by OwenKephart** [comment]
Hi @andrew-arkhipov, sorry for the delay (we were out on an offsite last week). This has just been merged in and will go out in this week's release (0.14.4).

I believe the pickling/unpickling issue that you were seeing was caused by having different versions of dagster running on the host machine vs in databricks, but this should be resolved by just using the latest release once it goes out. I ran some sample jobs where I uploaded the egg file (`python setup.py bdist_egg` in python_modules/dagster) directly to a databricks cluster and that seemed to work without issue.
